File: src/Backend.cpp
#include "Backend.h"

#include "Paths.h"
#include "Search.h"
#include "Thumbs.h"

#include <QDir>
#include <QFile>
#include <QFileDialog>
#include <QFileInfo>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QNetworkReply>
#include <QNetworkRequest>
#include <QProcess>
#include <QRandomGenerator>
#include <QRegularExpression>
#include <QStandardPaths>
#include <QTextStream>
#include <QTime>
#include <thread>

namespace {
const QStringList TRANSITIONS = {"simple", "fade",   "left",   "right", "top",    "bottom",
                                 "wipe",   "grow",   "center", "outer", "random", "wave"};

QString cacheRootStr() {
    return QDir::home().filePath(".cache/wallpaper_picker");
}

QNetworkRequest netReq(const QUrl &url) {
    QNetworkRequest r(url);
    r.setHeader(QNetworkRequest::UserAgentHeader,
                QStringLiteral("Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
                               "(KHTML, like Gecko) Chrome/126.0 Safari/537.36"));
    r.setTransferTimeout(90000);
    return r;
}

QString splitCmd(const QString &cmd) {
    // keep it simple: splitCommand from Qt (handles quotes)
    const auto parts = QProcess::splitCommand(cmd);
    if (parts.isEmpty())
        return {};
    return parts.value(0);
}
} // namespace

Backend::Backend(QObject *parent) : QObject(parent) {
    m_nam = new QNetworkAccessManager(this);
    m_search = new Search(this);
    connect(m_search, &Search::errorOccurred, this,
            [this](const QString &e) { queueNotice(tr("Búsqueda: %1").arg(e)); });

    reloadConfig();

    const QString envDir = qEnvironmentVariable("WALLPAPER_DIR");
    m_wallpaperDir = !envDir.isEmpty() ? envDir
                                       : configValue("wallpaper_dir",
                                                     QDir::home().filePath("Pictures/Wallpapers"));

    m_noticeTimer.setSingleShot(true);
    m_noticeTimer.setInterval(3500);
    connect(&m_noticeTimer, &QTimer::timeout, this, &Backend::clearNotice);

    connect(&m_watcher, &QFileSystemWatcher::directoryChanged, this, [this](const QString &path) {
        const QString t = QDir(cacheRootStr()).filePath("thumbs");
        const QString s = QDir(cacheRootStr()).filePath("search_thumbs");
        const QString m = QDir(cacheRootStr()).filePath("colors_markers");
        if (path == t)
            emit thumbsRefreshed();
        else if (path == s)
            emit searchRefreshed();
        else if (path == m)
            emit markersRefreshed();
        else if (QFileInfo(path).absoluteFilePath() == QFileInfo(m_wallpaperDir).absoluteFilePath())
            emit srcRefreshed();
    });
}

QString Backend::homeDir() const { return "file://" + QDir::homePath(); }
QString Backend::thumbDir() const { return "file://" + cacheRootStr() + "/thumbs"; }
QString Backend::searchDir() const { return "file://" + cacheRootStr() + "/search_thumbs"; }
QString Backend::markerDir() const { return "file://" + cacheRootStr() + "/colors_markers"; }

void Backend::reloadConfig() {
    m_cfg.clear();
    QFile f(Paths::configPathStr());
    if (!f.open(QIODevice::ReadOnly))
        return;
    for (const QString &line : QString::fromUtf8(f.readAll()).split('\n')) {
        const QString l = line.trimmed();
        if (l.isEmpty() || l.startsWith('#') || !l.contains('='))
            continue;
        const QString k = l.section('=', 0, 0).trimmed();
        const QString v = l.section('=', 1).trimmed();
        m_cfg.insert(k, v);
    }
}

QString Backend::configValue(const QString &key, const QString &fallback) const {
    return m_cfg.value(key, fallback);
}

void Backend::setNotice(const QString &msg) {
    m_notice = msg;
    emit noticeChanged();
    m_noticeTimer.start();
}

void Backend::queueNotice(const QString &msg) {
    QMetaObject::invokeMethod(this, "setNotice", Qt::QueuedConnection, Q_ARG(QString, msg));
}

void Backend::clearNotice() {
    m_notice.clear();
    emit noticeChanged();
}

void Backend::setDownloading(bool value, const QString &name) {
    m_downloading = value;
    m_downloadingName = name;
    emit downloadStateChanged();
}

void Backend::queueDownloading(bool value, const QString &name) {
    QMetaObject::invokeMethod(this, "setDownloading", Qt::QueuedConnection, Q_ARG(bool, value),
                              Q_ARG(QString, name));
}

void Backend::ensureDirs() {
    const QStringList dirs = {QDir(cacheRootStr()).filePath("thumbs"),
                              QDir(cacheRootStr()).filePath("search_thumbs"),
                              QDir(cacheRootStr()).filePath("colors_markers"), m_wallpaperDir};
    for (const QString &d : dirs)
        QDir().mkpath(d);
    setupWatchers();
}

void Backend::setupWatchers() {
    const QStringList dirs = {QDir(cacheRootStr()).filePath("thumbs"),
                              QDir(cacheRootStr()).filePath("search_thumbs"),
                              QDir(cacheRootStr()).filePath("colors_markers"), m_wallpaperDir};
    for (const QString &d : dirs)
        m_watcher.addPath(d);
}

void Backend::persistConfig() {
    QFile f(Paths::configPathStr());
    QDir().mkpath(QFileInfo(f).absolutePath());
    QStringList lines;
    if (f.open(QIODevice::ReadOnly))
        lines = QString::fromUtf8(f.readAll()).split('\n');
    f.close();
    QStringList out;
    bool found = false;
    for (const QString &line : lines) {
        if (line.trimmed().startsWith("wallpaper_dir")) {
            out << "wallpaper_dir=" + m_wallpaperDir;
            found = true;
        } else {
            out << line;
        }
    }
    if (!found)
        out << "wallpaper_dir=" + m_wallpaperDir;
    if (f.open(QIODevice::WriteOnly | QIODevice::Truncate))
        f.write(out.join('\n').toUtf8());
}

void Backend::resetLocalCache() {
    QDir(QDir(cacheRootStr()).filePath("thumbs")).removeRecursively();
    QDir(QDir(cacheRootStr()).filePath("colors_markers")).removeRecursively();
    QDir().mkpath(QDir(cacheRootStr()).filePath("thumbs"));
    QDir().mkpath(QDir(cacheRootStr()).filePath("colors_markers"));
}

void Backend::chooseWallpaperDir() {
    const QString chosen = QFileDialog::getExistingDirectory(
        nullptr, QStringLiteral("Selecciona la carpeta de wallpapers"), m_wallpaperDir);
    if (chosen.isEmpty())
        return;
    if (QFileInfo(chosen).absoluteFilePath() == QFileInfo(m_wallpaperDir).absoluteFilePath())
        return;
    m_wallpaperDir = QFileInfo(chosen).absoluteFilePath();
    persistConfig();
    resetLocalCache();
    setupWatchers();
    emit wallpaperDirChanged();
    emit localCacheReset();
    extractColors();
}

void Backend::loadMonitors() {
    const QStringList names = detectMonitors();
    QVariantList list;
    for (const QString &n : names)
        list << n;
    emit monitorsReady(list);
}

QStringList Backend::detectMonitors() const {
    struct Cmd { QStringList argv; };
    const QList<Cmd> cmds = {
        {{"niri", "msg", "outputs"}},
        {{"hyprctl", "monitors", "-j"}},
        {{"swaymsg", "-t", "get_outputs", "-r"}},
    };
    QSet<QString> seen;
    QStringList names;
    for (const Cmd &c : cmds) {
        QProcess p;
        p.start(c.argv.value(0), c.argv.mid(1));
        if (!p.waitForStarted(2000))
            continue;
        if (!p.waitForFinished(2000) || p.exitStatus() != QProcess::NormalExit)
            continue;
        const QByteArray out = p.readAllStandardOutput().trimmed();
        if (out.isEmpty())
            continue;
        const QJsonDocument doc = QJsonDocument::fromJson(out);
        if (!doc.isArray())
            continue;
        for (const QJsonValue &v : doc.array()) {
            const QString name = v.toObject().value("name").toString();
            if (!name.isEmpty() && !seen.contains(name)) {
                seen.insert(name);
                names << name;
            }
        }
        if (!names.isEmpty())
            break;
    }
    return names;
}

void Backend::applyWallpaper(const QString &safeName, bool isVideo, const QString &outputs) {
    const QString local = QDir(m_wallpaperDir).filePath(safeName);
    if (QFileInfo::exists(local)) {
        const QString name = safeName;
        if (isVideo) {
            applyVideo(local, outputs);
        } else {
            applyImage(local, outputs);
        }
        runWal(isVideo ? QDir(cacheRootStr()).filePath("thumbs/" + safeName) : local);
        emit wallpaperApplied(name);
        return;
    }
    const QString url = mapUrl(safeName);
    if (url.isEmpty()) {
        queueNotice(tr("No se encuentra el archivo: %1").arg(safeName));
        return;
    }
    m_pendingApplyName = safeName;
    m_pendingApplyVideo = isVideo;
    m_pendingApplyOutputs = outputs;
    setDownloading(true, safeName);
    downloadFull(safeName, url);
}

QString Backend::mapUrl(const QString &name) const {
    QFile f(QDir(cacheRootStr()).filePath("search_map.txt"));
    if (!f.open(QIODevice::ReadOnly))
        return {};
    for (const QString &line : QString::fromUtf8(f.readAll()).split('\n')) {
        const QString fname = line.section('|', 0, 0).trimmed();
        if (fname == name.trimmed())
            return line.section('|', 1).trimmed();
    }
    return {};
}

void Backend::downloadFull(const QString &safeName, const QString &url) {
    if (m_downloadReply)
        m_downloadReply->abort();
    auto *reply = m_nam->get(netReq(QUrl(url)));
    m_downloadReply = reply;
    connect(reply, &QNetworkReply::finished, this, [this, reply, safeName] {
        if (reply == m_downloadReply)
            m_downloadReply = nullptr;
        if (reply->error() != QNetworkReply::NoError) {
            setDownloading(false);
            queueNotice(tr("Descarga fallida: %1").arg(reply->errorString()));
            reply->deleteLater();
            return;
        }
        const QByteArray data = reply->readAll();
        reply->deleteLater();

        const QString target = QDir(m_wallpaperDir).filePath(safeName);
        const QString tmp = target + ".tmp";
        QFile out(tmp);
        if (!out.open(QIODevice::WriteOnly)) {
            setDownloading(false);
            queueNotice(tr("Descarga fallida: no se pudo crear el archivo"));
            return;
        }
        if (data.startsWith("RIFF") && data.mid(8, 4) == "WEBP") {
            QImage img;
            img.loadFromData(data);
            if (img.isNull()) {
                out.write(data);
            } else {
                QImage rgb = img.convertToFormat(QImage::Format_RGB32);
                rgb.save(&out, "JPEG", 92);
            }
        } else {
            out.write(data);
        }
        out.close();
        QFile::remove(target);
        QFile::rename(tmp, target);

        thumbs::makeThumb(target, QDir(cacheRootStr()).filePath("thumbs/" + safeName));
        setDownloading(false);

        const QString name = safeName;
        if (m_pendingApplyVideo) {
            applyVideo(target, m_pendingApplyOutputs);
        } else {
            applyImage(target, m_pendingApplyOutputs);
        }
        runWal(m_pendingApplyVideo ? QDir(cacheRootStr()).filePath("thumbs/" + safeName) : target);
        emit wallpaperApplied(name);
    });
}

bool Backend::applyImage(const QString &path, const QString &outputs) {
    const Setter st = detectSetter();
    QStringList cmd;
    if (st.mode == "custom") {
        cmd = QProcess::splitCommand(configValue("wpaper_cmd").replace(QStringLiteral("%s"), path));
    } else if (st.mode == "sway" || st.mode == "feh" || st.mode == "nitrogen") {
        pkill(st.mode == "sway" ? "swaybg" : st.mode);
        if (st.mode == "sway")
            cmd = {st.exe, "-i", path, "-m", "fill"};
        else if (st.mode == "feh")
            cmd = {st.exe, "--bg-fill", path};
        else
            cmd = {st.exe, "--set-scaled", path};
    } else { // awww / swww
        const QString transition =
            TRANSITIONS.at(QRandomGenerator::global()->bounded(TRANSITIONS.size()));
        cmd = {st.exe, "img", path, "--transition-type", transition,
               "--transition-pos", "0.5,0.5", "--transition-fps", "144", "--transition-duration",
               "1"};
        if (outputs != "all" && !outputs.isEmpty())
            cmd << "-o" << outputs;
    }
    if (cmd.isEmpty()) {
        queueNotice(tr("No se detectó ningún setter (awww/swww/swaybg/feh/nitrogen)"));
        return false;
    }
    QProcess::startDetached(cmd.value(0), cmd.mid(1));
    QFile::remove(QDir(cacheRootStr()).filePath("current_wallpaper.png"));
    QFile::copy(path, QDir(cacheRootStr()).filePath("current_wallpaper.png"));
    return true;
}

bool Backend::applyVideo(const QString &path, const QString &outputs) {
    const QString mpvpaper = QStandardPaths::findExecutable("mpvpaper");
    if (mpvpaper.isEmpty()) {
        queueNotice(tr("mpvpaper no está instalado"));
        return false;
    }
    const QString opts =
        "loop --no-audio --hwdec=auto --video-sync=display-resample --interpolation --tscale=oversample";
    QStringList targets;
    if (outputs.isEmpty() || outputs == "all")
        targets << "*";
    else
        targets = outputs.split(',');
    for (const QString &mon : targets)
        QProcess::startDetached(mpvpaper, {"-o", opts, mon, path});
    return true;
}

void Backend::runWal(const QString &img) {
    QString wal = QStandardPaths::findExecutable("wal");
    QStringList cmd;
    const QString userCmd = configValue("wal_cmd");
    if (!userCmd.isEmpty()) {
        QString c = userCmd;
        cmd = QProcess::splitCommand(c.replace(QStringLiteral("%s"), img));
    } else if (!wal.isEmpty())
        cmd = {wal, "-i", img, "-q"};
    if (cmd.isEmpty())
        return;

    const QString logPath = QDir(cacheRootStr()).filePath("wal.log");
    auto *p = new QProcess(this);
    p->setStandardOutputFile(logPath, QIODevice::Append);
    p->setStandardErrorFile(logPath, QIODevice::Append);
    p->setProcessChannelMode(QProcess::SeparateChannels);
    connect(p, QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished), this,
            [this, p](int, QProcess::ExitStatus) { p->deleteLater(); });
    {
        QFile f(logPath);
        if (f.open(QIODevice::Append)) {
            QTextStream ts(&f);
            ts << "\n[" << QTime::currentTime().toString("HH:mm:ss") << "] "
               << cmd.join(' ') << "\n";
        }
    }
    p->start(cmd.value(0), cmd.mid(1));
}

Backend::Setter Backend::detectSetter() const {
    const QString wanted = configValue("setter", "auto").toLower();
    if (!configValue("wpaper_cmd").isEmpty())
        return {QString(), "custom"};

    QStringList order;
    if (wanted != "auto")
        order << wanted;
    else
        order << QStringList{"awww", "swww", "swaybg", "feh", "nitrogen"};

    for (const QString &name : order) {
        const QString exe = QStandardPaths::findExecutable(name);
        if (exe.isEmpty())
            continue;
        if (name == "awww" || name == "swww") {
            if (ensureDaemon(exe))
                return {exe, "awww"};
        } else if (name == "swaybg") {
            return {exe, "sway"};
        } else if (name == "feh") {
            return {exe, "feh"};
        } else if (name == "nitrogen") {
            return {exe, "nitrogen"};
        }
    }
    return {QString(), QString()};
}

bool Backend::ensureDaemon(const QString &exe) const {
    QProcess pgrep;
    pgrep.start("pgrep", {"-f", "awww-daemon|swww-daemon"});
    if (pgrep.waitForStarted(2000) && pgrep.waitForFinished(2000))
        if (pgrep.exitCode() == 0)
            return true;
    QProcess::startDetached(exe, {"init"});
    return true;
}

void Backend::pkill(const QString &name) {
    if (name.isEmpty())
        return;
    QProcess::startDetached("pkill", {"-f", name});
}

void Backend::setSearchPaused(bool paused) { m_search->setPaused(paused); }

void Backend::search(const QString &query) {
    if (query.trimmed().isEmpty())
        return;
    m_search->start(query, QDir(cacheRootStr()).filePath("search_thumbs"),
                    QDir(cacheRootStr()).filePath("search_map.txt"));
}

void Backend::extractColors() {
    if (m_extracting.exchange(true))
        return;
    const QString src = m_wallpaperDir;
    const QString thumbsDir = QDir(cacheRootStr()).filePath("thumbs");
    const QString markersDir = QDir(cacheRootStr()).filePath("colors_markers");
    std::thread([this, src, thumbsDir, markersDir] {
        thumbs::extractWork(src, thumbsDir, markersDir);
        m_extracting.store(false);
    }).detach();
}

void Backend::shutdown() {
    m_search->stop();
    if (m_downloadReply)
        m_downloadReply->abort();
    clearNotice();
}

File: python-legacy/backend.py
import io
import json
import os
import shlex
import shutil
import subprocess
import threading
import time
from pathlib import Path

from PIL import Image
from PySide6.QtCore import (QFileSystemWatcher, QMetaObject, QObject, Qt,
                            QTimer, Property, Signal, Slot)
from PySide6.QtWidgets import QFileDialog

from search import fetch, run_search_thread
from thumbs import make_thumb, work_extract

HOME = Path(os.path.expanduser("~"))
CACHE_ROOT = HOME / ".cache" / "wallpaper_picker"
CONFIG_PATH = HOME / ".config" / "wallpaper_picker" / "config.conf"
DEFAULT_DIR = HOME / "Pictures" / "Wallpapers"

TRANSITIONS = ["simple", "fade", "left", "right", "top", "bottom", "wipe",
               "grow", "center", "outer", "random", "wave"]


class Config:
    def __init__(self, path=None):
        self.raw = {}
        p = Path(path) if path else Path(CONFIG_PATH)
        try:
            for line in p.read_text(errors="ignore").splitlines():
                line = line.strip()
                if not line or line.startswith("#") or "=" not in line:
                    continue
                k, _, v = line.partition("=")
                self.raw[k.strip()] = v.strip()
        except Exception:
            pass

    def get(self, key, default=""):
        return self.raw.get(key, default)


class Paths(QObject):
    @Property(str, constant=True)
    def logDir(self):
        return str(CACHE_ROOT)

    @Slot(str)
    def getCacheDir(self, name):
        return str(HOME / ".cache" / name)

    @Slot(str)
    def getRunDir(self, name):
        return str(CACHE_ROOT / name)


class Backend(QObject):
    monitorsReady = Signal(list)
    wallpaperApplied = Signal(str)
    noticeChanged = Signal()
    downloadStateChanged = Signal()
    wallpaperDirChanged = Signal()
    localCacheReset = Signal()
    thumbsRefreshed = Signal()
    searchRefreshed = Signal()
    markersRefreshed = Signal()
    srcRefreshed = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self.cfg = Config()
        wd = os.environ.get("WALLPAPER_DIR") or self.cfg.get("wallpaper_dir") or str(DEFAULT_DIR)
        self._wallpaper_dir = str(Path(wd).expanduser())
        self._notice = ""
        self._downloading = False
        self._downloading_name = ""
        self._setter = None
        self._stop_event = threading.Event()
        self._pause_event = threading.Event()
        self._search_thread = None
        self._extracting = False
        self._notice_timer = QTimer(self)
        self._notice_timer.setSingleShot(True)
        self._notice_timer.timeout.connect(self._clear_notice)

        self._watcher = QFileSystemWatcher(self)
        self._watcher.directoryChanged.connect(self._on_dir_changed)

    # ------------------------------------------------------------------ props
    @Property(str, notify=noticeChanged)
    def notice(self):
        return self._notice

    @Property(bool, notify=downloadStateChanged)
    def isDownloading(self):
        return self._downloading

    @Property(str, notify=downloadStateChanged)
    def downloadingName(self):
        return self._downloading_name

    @Property(str, constant=True)
    def homeDir(self):
        return "file://" + str(HOME)

    @Property(str, constant=True)
    def thumbDir(self):
        return "file://" + str(CACHE_ROOT / "thumbs")

    @Property(str, constant=True)
    def searchDir(self):
        return "file://" + str(CACHE_ROOT / "search_thumbs")

    @Property(str, notify=wallpaperDirChanged)
    def wallpaperDir(self):
        return self._wallpaper_dir

    def _set_notice(self, msg):
        self._notice = msg
        self.noticeChanged.emit()
        QMetaObject.invokeMethod(self, "_schedule_clear", Qt.ConnectionType.QueuedConnection)

    @Slot()
    def _schedule_clear(self):
        self._notice_timer.start(3500)

    def _clear_notice(self):
        self._notice = ""
        self.noticeChanged.emit()

    def _set_downloading(self, name="", value=False):
        self._downloading = value
        self._downloading_name = name
        self.downloadStateChanged.emit()

    # ---------------------------------------------------------- dirs & config
    @Slot()
    def ensureDirs(self):
        for d in (CACHE_ROOT / "thumbs", CACHE_ROOT / "search_thumbs",
                  CACHE_ROOT / "colors_markers", Path(self._wallpaper_dir)):
            try:
                d.mkdir(parents=True, exist_ok=True)
            except Exception:
                pass
        self._setup_watchers()

    def _watch_dirs(self):
        return {str(v): sig for v, sig in [
            (CACHE_ROOT / "thumbs", self.thumbsRefreshed),
            (CACHE_ROOT / "search_thumbs", self.searchRefreshed),
            (CACHE_ROOT / "colors_markers", self.markersRefreshed),
            (Path(self._wallpaper_dir), self.srcRefreshed),
        ]}

    def _setup_watchers(self):
        watched = self._watcher.directories()
        for p in self._watch_dirs():
            if p not in watched:
                self._watcher.addPath(p)

    def _on_dir_changed(self, path):
        sig = self._watch_dirs().get(path)
        if sig is not None:
            sig.emit()

    # ------------------------------------------------------- wallpaper dir
    @Slot()
    def chooseWallpaperDir(self):
        chosen = QFileDialog.getExistingDirectory(None, "Selecciona la carpeta de wallpapers",
                                                  self._wallpaper_dir)
        if not chosen:
            return
        chosen = str(Path(chosen).expanduser())
        if Path(chosen) == Path(self._wallpaper_dir):
            return
        self._wallpaper_dir = chosen
        self._persist_config()
        self._reset_local_cache()
        self._setup_watchers()
        self.wallpaperDirChanged.emit()
        self.localCacheReset.emit()
        self.extractColors()

    def _persist_config(self):
        try:
            p = Path(CONFIG_PATH)
            p.parent.mkdir(parents=True, exist_ok=True)
            lines = p.read_text(errors="ignore").splitlines() if p.exists() else []
            out, found = [], False
            for line in lines:
                if line.strip().startswith("wallpaper_dir"):
                    out.append("wallpaper_dir=%s" % self._wallpaper_dir)
                    found = True
                else:
                    out.append(line)
            if not found:
                out.append("wallpaper_dir=%s" % self._wallpaper_dir)
            p.write_text("\n".join(out) + "\n")
        except Exception:
            pass

    def _reset_local_cache(self):
        for d in ("thumbs", "colors_markers"):
            shutil.rmtree(str(CACHE_ROOT / d), ignore_errors=True)
        (CACHE_ROOT / "thumbs").mkdir(parents=True, exist_ok=True)
        (CACHE_ROOT / "colors_markers").mkdir(parents=True, exist_ok=True)

    # ---------------------------------------------------------------- monitors
    @Slot()
    def loadMonitors(self):
        names = self._monitors()
        self.monitorsReady.emit(names)

    def _monitors(self):
        candidates = [
            (["niri", "msg", "outputs"], lambda d: [o.get("name") for o in d if o.get("name")]),
            (["hyprctl", "monitors", "-j"], lambda d: [o.get("name") for o in d if o.get("name")]),
            (["swaymsg", "-t", "get_outputs", "-r"],
             lambda d: [o.get("name") for o in d if o.get("name") and o.get("active")]),
        ]
        for argv, proj in candidates:
            try:
                r = subprocess.run(argv, capture_output=True, text=True, timeout=3)
                if r.returncode != 0:
                    continue
                names = [n for n in proj(json.loads(r.stdout)) if n]
                if names:
                    return names
            except Exception:
                continue
        return []

    # --------------------------------------------------------------- wallpaper
    @Slot(str, bool, str)
    def applyWallpaper(self, safeName, isVideo, outputs):
        threading.Thread(target=self._apply_worker, args=(safeName, isVideo, outputs),
                         daemon=True).start()

    def _apply_worker(self, safeName, isVideo, outputs):
        try:
            src = Path(self._wallpaper_dir) / safeName
            from_search = not src.is_file()
            if from_search:
                url = self._map_url(safeName)
                if not url:
                    self._set_notice("No se encuentra el archivo: " + safeName)
                    return
                self._set_downloading(safeName, True)
                try:
                    src = self._download_search_image(safeName, url)
                except Exception as e:
                    self._set_notice("Descarga fallida: %s" % e)
                    self._set_downloading()
                    return
                finally:
                    if self._downloading:
                        self._set_downloading()

            self._run_setter(src, outputs, isVideo)

            try:
                shutil.copyfile(src, CACHE_ROOT / "current_wallpaper.png")
            except Exception:
                pass

            if isVideo:
                thumb = CACHE_ROOT / "thumbs" / ("000_" + src.name)
                wal_img = thumb if thumb.exists() else src
            else:
                wal_img = src
                self._maybe_thumb(src, safeName)
            self._run_wal(wal_img)

            reload_cmd = self.cfg.get("reload_cmd")
            if reload_cmd:
                subprocess.Popen(reload_cmd, shell=True,
                                 stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            self.wallpaperApplied.emit(safeName)
        except Exception as e:
            self._set_notice("No se pudo aplicar: %s" % e)

    def _maybe_thumb(self, src, safeName):
        thumb = CACHE_ROOT / "thumbs" / safeName
        try:
            if not thumb.exists():
                make_thumb(src, thumb)
        except Exception:
            pass

    def _map_url(self, name):
        try:
            mf = CACHE_ROOT / "search_map.txt"
            for line in mf.read_text(errors="ignore").splitlines():
                fname, _, url = line.partition("|")
                if fname.strip() == name.strip():
                    return url.strip()
        except Exception:
            pass
        return ""

    def _download_search_image(self, safeName, url):
        src = Path(self._wallpaper_dir) / safeName
        tmp = src.with_name(src.name + ".tmp")
        data = fetch(url, timeout=90)
        if data[:4] == b"RIFF" and data[8:12] == b"WEBP":
            im = Image.open(io.BytesIO(data)).convert("RGB")
            im.save(tmp, "JPEG", quality=92)
        else:
            tmp.write_bytes(data)
        os.replace(tmp, src)
        thumb = CACHE_ROOT / "thumbs" / safeName
        try:
            if not thumb.exists():
                make_thumb(src, thumb)
        except Exception:
            pass
        return src

    def _run_setter(self, img, outputs, isVideo):
        if isVideo:
            self._apply_video(img, outputs)
            return
        setter, mode = self._detect_setter()

        if mode == "custom":
            cmd = shlex.split(self.cfg.get("wpaper_cmd").replace("%s", shlex.quote(str(img))))
            subprocess.Popen(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            return

        if setter is None:
            raise RuntimeError("no se detectó ningún setter (awww/swww/swaybg/feh/nitrogen)")

        if mode == "sway":
            self._pkill("swaybg")
            cmd = [setter, "-i", str(img), "-m", "fill"]
        elif mode == "feh":
            self._pkill("feh")
            cmd = [setter, "--bg-fill", str(img)]
        elif mode == "nitrogen":
            self._pkill("nitrogen")
            cmd = [setter, "--set-scaled", str(img)]
        else:  # awww / swww
            if outputs and outputs != "all":
                cmd = self._awww_args(setter, img) + ["-o", outputs]
            else:
                cmd = self._awww_args(setter, img)

        subprocess.Popen(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

    def _awww_args(self, setter, img):
        t = TRANSITIONS[int(time.time() * 1000) % len(TRANSITIONS)]
        return [setter, "img", str(img), "--transition-type", t,
                "--transition-pos", "0.5,0.5", "--transition-fps", "144",
                "--transition-duration", "1"]

    def _apply_video(self, img, outputs):
        mpvpaper = shutil.which("mpvpaper")
        if not mpvpaper:
            raise RuntimeError("mpvpaper no está instalado")
        opts = "loop --no-audio --hwdec=auto --video-sync=display-resample " \
               "--interpolation --tscale=oversample"
        targets = ["*"] if (not outputs or outputs == "all") else outputs.split(",")
        for mon in targets:
            subprocess.Popen([mpvpaper, "-o", opts, mon, str(img)],
                             stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

    def _pkill(self, name):
        try:
            subprocess.run(["pkill", "-f", name], capture_output=True)
        except Exception:
            pass

    def _detect_setter(self):
        wanted = self.cfg.get("setter", "auto").lower()
        if self.cfg.get("wpaper_cmd"):
            return None, "custom"

        order = []
        if wanted and wanted != "auto":
            order.append(wanted)
        else:
            order.extend(["awww", "swww", "swaybg", "feh", "nitrogen"])

        for name in order:
            exe = shutil.which(name)
            if not exe:
                continue
            if name in ("awww", "swww"):
                if self._daemon_running() or self._try_init(exe):
                    return exe, "awww"
            elif name == "swaybg":
                return exe, "sway"
            elif name == "feh":
                return exe, "feh"
            elif name == "nitrogen":
                return exe, "nitrogen"
        return None, None

    def _daemon_running(self):
        try:
            r = subprocess.run(["pgrep", "-f", "awww-daemon|swww-daemon"],
                               capture_output=True, text=True)
            return r.returncode == 0
        except Exception:
            return False

    def _try_init(self, exe):
        try:
            r = subprocess.run([exe, "init"], capture_output=True, timeout=10)
            return r.returncode == 0
        except Exception:
            return False

    def _run_wal(self, img):
        wal = shutil.which("wal")
        if not wal:
            return
        user_cmd = self.cfg.get("wal_cmd")
        try:
            if user_cmd:
                cmd = shlex.split(user_cmd.replace("%s", shlex.quote(str(img))))
            else:
                cmd = [wal, "-i", str(img), "-q"]
            log = CACHE_ROOT / "wal.log"
            with open(log, "a") as f:
                f.write("\n[%s] %s\n" % (time.strftime("%H:%M:%S"), " ".join(cmd)))
                subprocess.Popen(cmd, stdout=f, stderr=f)
        except Exception:
            pass

    # ------------------------------------------------------------------ search
    @Slot(bool)
    def setSearchPaused(self, paused):
        if paused:
            self._pause_event.set()
        else:
            self._pause_event.clear()

    @Slot(str)
    def search(self, query):
        if not (query or "").strip():
            return
        self._pause_event.clear()
        self._stop_event.clear()
        self._search_thread = run_search_thread(
            query, str(CACHE_ROOT / "search_thumbs"),
            str(CACHE_ROOT / "search_map.txt"),
            self._pause_event, self._stop_event,
            lambda e: self._set_notice("Búsqueda: " + e))

    # ----------------------------------------------------------------- thumbs
    @Slot()
    def extractColors(self):
        if self._extracting:
            return
        self._extracting = True
        threading.Thread(target=self._extract_worker, daemon=True).start()

    def _extract_worker(self):
        try:
            work_extract(self._wallpaper_dir, CACHE_ROOT / "thumbs",
                         CACHE_ROOT / "colors_markers")
        finally:
            self._extracting = False

    @Slot()
    def shutdown(self):
        self._stop_event.set()
        self._set_notice("")
